docs: revise per feedback from @blipp and @grittygrease (#8)
cfm/draft-ms-hpke-auth-modes: draft-ms-hpke-auth-modes.md

a post-quantum KEM shared secret as the PSK to achieve hybrid PQ/T
confidentiality. That material is provided to motivate the extension
and may be developed as separate work.

Implicit authentication does not provide non-repudiation, and the
authenticated modes inherit the Key Compromise Impersonation (KCI)
vulnerability common to all DH-based authentication schemes; both
limitations are discussed in {{sec-security}}.

# Conventions and Definitions
{::boilerplate bcp14-tagged}

Terms from {{I-D.ietf-hpke-hpke}} are used without redefinition. The
following additional term is used herein:

- **AKEM**: a KEM whose encapsulation additionally takes the sender's
  static private key and implicitly binds the shared secret to it.

# Authenticated Mode Extensions to {{I-D.ietf-hpke-hpke}} {#sec-ext}

This section specifies the additions to {{!I-D.ietf-hpke-hpke}} required
to restore `mode_auth` and `mode_auth_psk`. These extensions
are defined so that the externally observable behavior of existing HPKE
modes is unchanged, although this document updates the `VerifyPSKInputs`
procedure in {{I-D.ietf-hpke-hpke}}.

## Mode Identifiers

> Two mode identifiers are added to the table in {{Section 5 of !I-D.ietf-hpke-hpke}}:

Reviewer: technically it's un-reserving those values, so maybe: 

```suggestion
The reserved entries for values 0x02 and 0x03 in Table 1 of {{!I-D.ietf-hpke-hpke}} are replaced with the following mode identifiers: 
```

~~~
mode_auth     = 0x02
mode_auth_psk = 0x03
~~~

## DHKEM Extension: AuthEncap and AuthDecap {#sec-authencap}

The following two functions are added to the DHKEM, extending it to an
AKEM. All helper functions (`GenerateKeyPair`, `DH`, `SerializePublicKey`,
`DeserializePublicKey`, `ExtractAndExpand`) are as defined in
{{I-D.ietf-hpke-hpke}}.

~~~
def AuthEncap(pkR, skS):
  skE, pkE = GenerateKeyPair()
  dh = concat(DH(skE, pkR), DH(skS, pkR))
  enc = SerializePublicKey(pkE)
  pkRm = SerializePublicKey(pkR)
  pkSm = SerializePublicKey(pk(skS))
  kem_context = concat(enc, pkRm, pkSm)
  shared_secret = ExtractAndExpand(dh, kem_context)
  return shared_secret, enc

def AuthDecap(enc, skR, pkS):
  pkE = DeserializePublicKey(enc)
  dh = concat(DH(skR, pkE), DH(skR, pkS))
  pkRm = SerializePublicKey(pk(skR))
  pkSm = SerializePublicKey(pkS)
  kem_context = concat(enc, pkRm, pkSm)
  shared_secret = ExtractAndExpand(dh, kem_context)
  return shared_secret
~~~

> 

Reviewer: perhaps worth adding a note about KCI here too (as RFC 9180 does):

```suggestion
Note that the AuthEncap() and AuthDecap() functions are vulnerable to key-compromise impersonation (KCI): the assurance that the shared secret was generated by the holder of the private key `skS` does not hold if the recipient private key `skR` is compromised. See {{sec-security}} for further discussion.
```

`[mode_base, mode_auth]` and `mode_psk` is replaced by
`[mode_psk, mode_auth_psk]` in the final two guards.

## Setup Functions {#sec-setup}

The following four Setup functions are added alongside those defined in
{{I-D.ietf-hpke-hpke}}. `KeyScheduleS`/`KeyScheduleR` and
`AuthEncap`/`AuthDecap` are as defined in {{!I-D.ietf-hpke-hpke}} and
{{sec-authencap}} respectively.

~~~
def SetupAuthS(pkR, info, skS):
  shared_secret, enc = AuthEncap(pkR, skS)
  return enc, KeyScheduleS(mode_auth, shared_secret, info,
                           default_psk, default_psk_id)

def SetupAuthR(enc, skR, info, pkS):
  shared_secret = AuthDecap(enc, skR, pkS)
  return KeyScheduleR(mode_auth, shared_secret, info,
                      default_psk, default_psk_id)

def SetupAuthPSKS(pkR, info, psk, psk_id, skS):
  shared_secret, enc = AuthEncap(pkR, skS)
  return enc, KeyScheduleS(mode_auth_psk, shared_secret, info,
                           psk, psk_id)

def SetupAuthPSKR(enc, skR, info, psk, psk_id, pkS):
  shared_secret = AuthDecap(enc, skR, pkS)
  return KeyScheduleR(mode_auth_psk, shared_secret, info,
                      psk, psk_id)
~~~

> 

Reviewer: so in RFC 9180, they have this section on validation of inputs and outputs: https://www.rfc-editor.org/rfc/rfc9180#name-validation-of-inputs-and-ou

as does the new HPKE draft: https://datatracker.ietf.org/doc/html/draft-ietf-hpke-hpke-03#name-validation-of-inputs-and-ou

comparing those, the authenticated modes have an additional requirement: 
> in authenticated modes, the recipient MUST validate the sender's static public key `pkS`. Validation failure yields a `ValidationError`

```suggestion
## Input Validation and Error Handling

In addition to the validation requirements in {{Section 7.1.4 of
!I-D.ietf-hpke-hpke}}, the recipient MUST validate the sender's
static public key `pkS` before use in `AuthDecap()`, applying the
same validation rules as for other public key inputs. Validation failure
MUST yield a `ValidationError`.
```

{{RFC5869}}. In `CombineSecrets_OneStage`, `dhkem_shared_secret` and
`psk` are length-prefixed and concatenated before a single
`LabeledDerive` call. In both cases, `dhkem_shared_secret` and `ss_pq`
enter the combination as independent inputs. The intended design
property is that `secret` remains pseudorandom as long as at least one
of the two inputs is — meaning an adversary would need to attack both
the classical DH-based component and the PQ-KEM to recover `secret`.
Whether this property holds formally for a specific `CombineSecrets`
variant depends on that variant's security analysis, which is outside
the scope of this document. Authentication remains entirely classical;
a quantum adversary that breaks the DH assumption can also forge sender
authentication, so post-quantum sender authentication would require an
additional PQ signature. Note that {{Alwen2023}} describes a related
hybrid construction in which a PQ *AKEM* (rather than a plain KEM) is
used to generate the PSK, which would additionally provide post-quantum
sender authentication; that stronger construction is outside the scope
of this document.

**PSK freshness.** The ML-KEM shared secret `ss_pq` satisfies the

> entropy requirement in {{Section 9.7 of !I-D.ietf-hpke-hpke}} (32 bytes

Reviewer: should be section 9.5? section 9.5 contains the entropy requirements:

> In the PSK modes, the PSK MUST have at least 32 bytes of entropy and SHOULD be of length Nh bytes or longer.

(though 9.7.5 discusses the impact of bad ephemeral randomness)
```suggestion
entropy requirement in {{Section 9.5 of !I-D.ietf-hpke-hpke}} (32 bytes
```
